Model/CreditmemoByIncrementId.php

```diff
@@ -0,0 +1,123 @@
+
+class CreditmemoByIncrementId implements CreditmemoByIncrementIdInterface
+{
+    /** @var CreditmemoManagementInterface */
+    private $creditmemoManagement;
+
+    /** @var CreditmemoRepositoryInterface */
+    private $creditmemoRepository;
+
+    /** @var OrderRepositoryInterface */
+    private $orderRepository;
+
+    /** @var CreditmemoFactory */
+    private $creditmemoFactory;
+
+    /**
+     * CreditmemoByIncrementId constructor.
+     * @param CreditmemoRepositoryInterface $creditmemoRepository
+     * @param CreditmemoManagementInterface $creditmemoManagement
+     * @param OrderRepositoryInterface $orderRepository
+     */
+    public function __construct(
+        CreditmemoRepositoryInterface $creditmemoRepository,
+        CreditmemoManagementInterface $creditmemoManagement,
+        OrderRepositoryInterface $orderRepository,
+        CreditmemoFactory $creditmemoFactory
+    ) {
+        $this->creditmemoRepository = $creditmemoRepository;
+        $this->creditmemoManagement = $creditmemoManagement;
+        $this->orderRepository = $orderRepository;
+        $this->creditmemoFactory = $creditmemoFactory;
+    }
+
+    /**
+     * Create a new creditmemo using increment_id and then create refund
+     *
+     * @param string $orderIncrementId
+     * @param CreditmemoInterface $creditmemo
+     * @return CreditmemoInterface
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function createAndRefund(
+        $orderIncrementId,
+        CreditmemoInterface $creditmemo
+    ) {
+        /** @var Order $order */
+        $order = $this->loadOrderByIncrementId($orderIncrementId);
+
+        if(!$order->canCreditmemo()) {
+            throw new \Magento\Framework\Exception\LocalizedException(
+                __("This order does not allow creation of creditmemo")
+            );
+        }
+
+        /** @var CreditmemoInterface $creditmemo */
+        $newCreditmemo = $this->creditmemoFactory->createByOrder($order, [
+            'qtys' => $this->filterItemsToBeRefunded($order, $creditmemo)
+        ]);
+        $newCreditmemo->setState(Creditmemo::STATE_OPEN);
+
+        $this->creditmemoRepository->save($newCreditmemo);
+
+        return $this->creditmemoManagement->refund($newCreditmemo);
+    }
+
+    /**
+     * Get the sku and qty from the input payload to decide which item and its quantity to be refunded.
+     * This allow partial creditmemo/refund
+     *
+     * @param Order $order
+     * @param CreditmemoInterface $creditmemo
+     * @return array
+     */
+    private function filterItemsToBeRefunded(Order $order, CreditmemoInterface $creditmemo)
+    {
+        $selectedItemsToRefund = [];
+        /** @var \Magento\Sales\Model\Order\Item $orderItem */
+        foreach($order->getAllItems() as $orderItem){
+            /** @var \Magento\Sales\Model\Order\Creditmemo\Item $inputItem */
+            foreach($creditmemo->getItems() as $inputItem){
```

Comment:
Is there a case where this is empty or null

Reply:
If it is null we might need to protect it otherwise 👍

Reply:
no it will be array.